jetty-server/src/main/java/org/eclipse/jetty/server/HttpConnection.java

```diff
@@ -885,8 +886,10 @@ private void releaseChunk()
         @Override
         protected void onCompleteSuccess()
         {
+            boolean upgrading = _info != null && _info.getStatus() == HttpStatus.SWITCHING_PROTOCOLS_101;
             release().succeeded();
-            if (_shutdownOut)
+            // If successfully upgraded it is responsibility of the next protocol to close the connection.
+            if (_shutdownOut && !upgrading)
```

Comment:
Perhaps this covers both cases:
```suggestion
            // If successfully upgraded it is responsibility of the next protocol to close the connection.
            if (_shutdownOut && _channel.getRequest().getAttribute(UPGRADE_CONNECTION_ATTRIBUTE) == null)
)
```